kernel/work.c

```diff
@@ -0,0 +1,1030 @@
+		flag_set(&work->flags, K_WORK_RUNNING_BIT);
+	} else {
+		/* Caught the item before being invoked; complete the
+		 * cancellation now.
+		 */
+		finalize_cancel_locked(work);
+	}
+
+	k_spin_unlock(&lock, key);
+
+	return ret;
+}
+
+/* Work handler has been called and is about to go idle.
+ *
+ * If the work is being canceled this will notify anything waiting
+ * for the cancellation.
+ *
+ * Invoked by work queue thread.
+ * Takes and releases lock.
+ * Reschedules via finalize_cancel_locked
+ *
+ * @param work work that is in running state
+ */
+static inline void work_clear_running(struct k_work *work)
+{
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	/* Clear running */
+	flag_clear(&work->flags, K_WORK_RUNNING_BIT);
+
+	if (flag_test(&work->flags, K_WORK_CANCELING_BIT)) {
+		finalize_cancel_locked(work);
+	}
+
+	k_spin_unlock(&lock, key);
+}
+
+/* Loop executed by a work queue thread.
+ *
+ * @param workq_ptr pointer to the work queue structure
+ */
+static void work_queue_main(void *workq_ptr, void *p2, void *p3)
+{
+	struct k_work_q *queue = (struct k_work_q *)workq_ptr;
+
+	while (true) {
+		sys_snode_t *node;
+		struct k_work *work = NULL;
+		k_spinlock_key_t key = k_spin_lock(&lock);
+
+		/* Clear the record of processing any previous work, and check
+		 * for new work.
+		 */
+		node = sys_slist_get(&queue->pending);
+		if (node != NULL) {
+			/* Mark that there's some work active that's
+			 * not on the pending list.
+			 */
+			flag_set(&queue->flags, K_WORK_QUEUE_BUSY_BIT);
+			work = CONTAINER_OF(node, struct k_work, node);
+		} else if (flag_test_and_clear(&queue->flags,
+					       K_WORK_QUEUE_DRAIN_BIT)) {
+			/* Not busy and draining: move threads waiting for
+			 * drain to ready state.  The held spinlock inhibits
+			 * immediate reschedule; released threads get their
+			 * chance when this invokes z_sched_wait() below.
+			 *
+			 * We don't touch K_WORK_QUEUE_PLUGGABLE, so getting
+			 * here doesn't mean that the queue will allow new
+			 * submissions.
+			 */
+			(void)z_sched_wake_all(&queue->drainq, 1, NULL);
+		}
+
+		if (work == NULL) {
+			/* Nothing's had a chance to add work since we took
+			 * the lock, and we didn't find work nor got asked to
+			 * stop.  Just go to sleep: when something happens the
+			 * work thread will be woken and we can check again.
+			 */
+
+			(void)z_sched_wait(&lock, key, &queue->notifyq,
+					   K_FOREVER, NULL);
+			continue;
+		}
+
+		k_spin_unlock(&lock, key);
+
+		if (work != NULL) {
+			bool yield;
+			k_work_handler_t handler = work->handler;
+
+			__ASSERT_NO_MSG(handler != 0);
+
+			if (work_set_running(work, queue)) {
+				handler(work);
+				work_clear_running(work);
+			}
+
+			/* No longer referencing the work, so we can clear the
+			 * BUSY flag while we yield to prevent starving other
+			 * threads.
+			 */
+			key = k_spin_lock(&lock);
+			flag_clear(&queue->flags, K_WORK_QUEUE_BUSY_BIT);
+			yield = !flag_test(&queue->flags, K_WORK_QUEUE_NO_YIELD_BIT);
```

Comment:
It's necessary to read flags under lock.  Rather than read it once on entry and dedicate stack space for the cached value it made more sense to re-read it here.